enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/IdentityVerificationOtpService.java

```diff
@@ -133,6 +155,54 @@ public boolean isUserVerifiedUsingOtp(String processId) {
         return otp.getStatus() == OtpStatus.VERIFIED;
     }
 
+    private OtpVerifyResponse verifyPresenceCheck(final OnboardingProcessEntity process, final OtpVerifyResponse response) throws OnboardingProcessException {
+        final String processId = process.getId();
+        if (!identityVerificationConfig.isPresenceCheckEnabled()) {
+            logger.debug("Presence check is not enabled, Process ID: {}", processId);
+            return response;
+        }
+
+        logger.debug("Evaluating result of presence check, Process ID: {}", processId);
+
+        final IdentityVerificationEntity idVerification = identityVerificationRepository.findFirstByActivationIdOrderByTimestampCreatedDesc(process.getActivationId()).orElseThrow(() ->
+                new OnboardingProcessException("No IdentityVerification found for Process ID: " + processId));
+        final String errorDetail = idVerification.getErrorDetail();
+        final String rejectReason = idVerification.getRejectReason();
+
+        if (StringUtils.isNotBlank(errorDetail) || StringUtils.isNotBlank(rejectReason)) {
+            logger.info("SCA failed, IdentityVerification ID: {} of Process ID: {} contains errorDetail: {}, rejectReason: {} from previous step",
+                    idVerification.getId(), processId, errorDetail, rejectReason);
+
+            idVerification.setPhase(PRESENCE_CHECK);
+            idVerification.setStatus(IdentityVerificationStatus.NOT_INITIALIZED);
+            idVerification.setTimestampLastUpdated(new Date());
+            idVerification.setErrorDetail("");
+            idVerification.setRejectReason("");
+            identityVerificationRepository.save(idVerification);
+
+            logger.info("Switched to PRESENCE_CHECK/NOT_INITIALIZED, Process ID: {}", idVerification.getProcessId());
+
+            processLimitService.incrementErrorScore(process, OnboardingProcessError.ERROR_USER_VERIFICATION_OTP_FAILED);
+            final OnboardingStatus status = processLimitService.checkOnboardingProcessErrorLimits(process).getStatus();
+            response.setOnboardingStatus(status);
+            response.setVerified(false);
+
+            markVerificationOtpAsFailed(processId);
+        } else {
+            logger.debug("PRESENCE_CHECK without error or reject reason, Process ID: {}", idVerification.getProcessId());
+        }
+        return response;
+    }
+
+    private void markVerificationOtpAsFailed(String processId) throws OnboardingProcessException {
+        final OnboardingOtpEntity otp = onboardingOtpRepository.findLastOtp(processId, OtpType.USER_VERIFICATION).orElseThrow(() ->
+            new OnboardingProcessException("Onboarding OTP not found for process ID: " + processId));
+        otp.setStatus(OtpStatus.FAILED);
+        otp.setErrorDetail(OnboardingOtpEntity.ERROR_CANCELED);
```

Comment:
Mark otp as failed just to be sure not to re-use it. The question is whether to introduce a new error detail.

Reply:
`Onboarding OTP not found, Process ID: ` ?